### PR TITLE
ModCore: Expanded Mod Support (Custom Characters and Songs with 0 Code!)

### PR DESCRIPTION
**Pull request by EliteMasterEric**
This pull request changes the following:

* Characters (including animations and offsets) can now be loaded from a single JSON file rather than being hard-coded.
  * Currently only GF has been converted, but with this, mods may now freely add their own characters.
  * Note that this also allows mods to replace or merge with these JSONs to fix character skins.
  * At some point we're going to want to move all the characters from the massive ugly case structure in the `Characters` constructor into the JSON files.
* Fixes to ensure custom songs in mods can be added to Free Play (by appending to the `freeplaySonglist.txt` file) without crashing.
* Song offsets are now stored inside a `_meta.json` in the song folder rather than in the NAME of a `.offset` file. This was necessary to allow mods to override it.
* De-hardcoded the logic which handles a song's name. Relevant code now refers to either the song ID (the name in the file system, such as `milf`) or the song name (the readable name to display to the user, such as `M.I.L.F`). The readable name is included in the `_meta.json`.

As stated in the title, we can now create mods which add new characters and songs. No code needed.

**[Watch the demo](https://www.youtube.com/watch?v=1A5ChNEf5F4)**, with included commentary discussing PolyMod.

Attached is the mod I developed to test this functionality. It adds Improbable Outset to the freeplay menu for you to play. Download and build this branch, then unzip and place it in the mods folder like in the demo.
[trickyMod.zip](https://github.com/KadeDev/Kade-Engine/files/7174947/trickyMod.zip)

### CODE REVIEW

**Comment by ghost** [comment]
Ooh this is actually really cool!

**Review by somefnfplayerlol — COMMENTED** [review]
what happened with the html5 ci build?

File: source/Character.hx [hunk 1]
```diff
@@ -424,6 +414,47 @@ class Character extends FlxSprite
...
+
+		var tex:FlxAtlasFrames = Paths.getSparrowAtlas(data.asset, 'shared');
+		frames = tex;
+
+		for (anim in data.animations)
+		{
+			var frameRate = anim.frameRate == null ? 24 : anim.frameRate;
+			var looped = anim.looped == null ? false : anim.looped;
+			var flipX = anim.flipX == null ? false : anim.looped;
+			var flipY = anim.flipY == null ? false : anim.looped;
```
Comment (NeeEoo) [inline source/Character.hx]: Shouldn't these be
```haxe
var flipX = anim.flipX == null ? false : anim.flipX;
var flipY = anim.flipY == null ? false : anim.flipY;
```

**Comment by isakube** [comment]
Friday Night Funkin': Actually Moddable (holy shit) (it happened) (and it works too) (woah)

**Review by isakube — APPROVED** [review]

**Comment by EliteMasterEric** [comment]
> what happened with the html5 ci build?

It broke because my fork of Polymod didn't disable itself on HTML5 builds properly, I fixed it but I need to get this PR's code build actions to trigger again, probably by making another commit...

> 
> 
> Shouldn't these be
> 
> ```hx
> var flipX = anim.flipX == null ? false : anim.flipX;
> var flipY = anim.flipY == null ? false : anim.flipY;
> ```

...and I found something to fix! Perfect.

File: source/Character.hx (hunk 1 above)
Comment (EliteMasterEric) [reply source/Character.hx]: That is correct, I have resolved this issue just now.

**Comment by isakube** [comment]
enigma balls
Screen Shot 2021-09-17 at 3 55 45 PM
Screen Shot 2021-09-17 at 3 56 12 PM

**Comment by EliteMasterEric** [comment]
> enigma balls
> Screen Shot 2021-09-17 at 3 55 45 PM
> Screen Shot 2021-09-17 at 3 56 12 PM

I see you read the readme.

`modcore-1.7` is an old branch on my fork I haven't touched in a while. I was messing with adding 9-key there but it's broken now because of the other changes I made. You can check out... uhhh... *rapid typing sounds* `master` branch to test out modding.

**Comment by ghost** [comment]
Like you explained your username

**Comment by isakube** [comment]
> I see you read the readme.

yeah mhm i totally read the readme on the repo and didn't just download the default branch like a dumbass
edit: ok wait the fuck?? download code jut gives me modcore-1.7 and if i try to clone a specific branch it fails whyyyyyyyy

**Comment by EliteMasterEric** [comment]
> > I see you read the readme.
> 
> yeah mhm i totally read the readme on the repo and didn't just download the default branch like a dumbass
> edit: ok wait the fuck?? download code jut gives me modcore-1.7 and if i try to clone a specific branch it fails whyyyyyyyy

You have to either:

* Switch to the branch you want in the Github interface and download the ZIP there.
* Clone the repo using Git and check out the branch you want. For this you can either learn to use git from the command line, or use one of a wide variety of Git GUIs for it (I like GitKraken). The latter is far more convenient for your day-to-day, but learning the command line is important; having a good understanding of how version control works is an essential skill for beginner programmers looking to become more competent.

**Comment by isakube** [comment]
> * Switch to the branch you want in the Github interface and download the ZIP there.

yeah i DID and it still gives me modcore-1.7

**Comment by EliteMasterEric** [comment]
> > * Switch to the branch you want in the Github interface and download the ZIP there.
> 
> yeah i DID and it still gives me modcore-1.7

Run the following commands then:

```
git clone https://github.com/MasterEric/Enigma-Engine
cd ./Enigma-Engine
git checkout master
```

**Comment by isakube** [comment]
> > > * Switch to the branch you want in the Github interface and download the ZIP there.
> > 
> > 
> > yeah i DID and it still gives me modcore-1.7
> 
> Run the following commands then:
> 
> ```
> git clone https://github.com/MasterEric/Enigma-Engine
> cd ./Enigma-Engine
> git checkout master
> ```

coolio

**Comment by somefnfplayerlol** [comment]
I'm considering the next version of KE 2.0

**Pull request merged**